### PR TITLE
statefulset persistent volume claim resize

### PR DESCRIPTION
**Pull request by areller**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:

Today when resizing a persistent volume claim that's created by a stateful set template, by modifying the size in the template, the users gets back a generic error, saying that they're only allowed to modify number of replicas or statefulset spec template.
This PR allows users to also modify the request size in a PVC template of a statefulset, and it modifies the statefulset controller to be able to reconcile request size differences (i.e. patches the PVC if the size in the template changes)

In addition,

1. Those kind of changes aren't affected by rollback (i.e. if you rollback a PVC resize, it won't actually resize back)
2. If the PVC patch fails during statefulset reconciliation (e.g. the user tries to decrease the PVC size and the storage driver doesn't support that), all other changes to the statefulset are blocked from being reconciled, until the user addresses the other (e.g. increase the PVC size back up in the statefulset template)

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: pkg/controller/statefulset/stateful_set_control.go [hunk 1]
```diff
@@ -592,6 +598,51 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
...
+				ssc.podControl.recordUnavailableClaimEvent("resize", set, pod, claimName, err)
+				continue
+			}
+
+			resizing := false
+			if pvcActual.Spec.Resources.Requests.Storage().Equal(*pvc.Spec.Resources.Requests.Storage()) {
+				continue
+			} else {
+				resizing = true
+			}
```
Comment (jaypipes) [inline pkg/controller/statefulset/stateful_set_control.go:626]: unless I'm mistaken, you can just shorten this to:

```go
curSize := pvcActual.Spec.Resources.Requests.Storage()
newSize := pvc.Spec.Resources.Requests.Storage()
if curSize.Equal(*newSize) {
    continue
}
```

File: pkg/controller/statefulset/stateful_set_control.go [hunk 2]
```diff
@@ -592,6 +598,51 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
...
+			if err != nil {
+				err = fmt.Errorf("failed to resize PVC %s: %s", claimName, err)
+				errs = append(errs, err)
+				ssc.podControl.recordClaimEvent("resize", set, pod, pvcActual, err)
+				continue
+			}
+
+			if resizing {
+				ssc.podControl.recordClaimEvent("resize", set, pod, pvcActual, nil)
+			}
```
Comment (jaypipes) [inline pkg/controller/statefulset/stateful_set_control.go:639]: You can shorten this whole block to just this:

```go
if err != nil {
    err = fmt.Errorf(("failed to resize PVC %s", claimName, err)
    errs - append(errs, err)
}
ssc.podControl.recordClaimEvent("resize", set, pod, pvc, err)
```

}

File: pkg/controller/statefulset/stateful_set_control.go [hunk 3]
```diff
@@ -592,6 +598,51 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
...
+
+			resizing := false
+			if pvcActual.Spec.Resources.Requests.Storage().Equal(*pvc.Spec.Resources.Requests.Storage()) {
+				continue
+			} else {
+				resizing = true
+			}
+
+			patch := fmt.Sprintf(`{"spec": {"resources": {"requests": {"storage": "%s"}}}}`, pvc.Spec.Resources.Requests.Storage().String())
```
Comment (jaypipes) [inline pkg/controller/statefulset/stateful_set_control.go:628]: If you take my suggested code above, you can do:

```go
patch := fmt.Sprintf(`{"spec": {"resources": {"requests": {"storage": "%s"}}}}`, curSize)
```

**Comment by areller** [comment]
Thank you @jaypipes. these are good suggestions
I'm waiting for https://github.com/kubernetes/enhancements/pull/3412 to be reviewed/approved
this PR is a draft, I'm not sure yet if I'll use it or open a new one for the full implementation, but I'll keep your suggestions in mind

File: pkg/controller/statefulset/stateful_pod_control.go [hunk 4]
```diff
@@ -301,15 +308,22 @@ func (spc *StatefulPodControl) recordPodEvent(verb string, set *apps.StatefulSet
 // nil the generated event will have a reason of v1.EventTypeNormal. If err is not nil the generated event will have a
 // reason of v1.EventTypeWarning.
 func (spc *StatefulPodControl) recordClaimEvent(verb string, set *apps.StatefulSet, pod *v1.Pod, claim *v1.PersistentVolumeClaim, err error) {
+	spc.recordUnavailableClaimEvent(verb, set, pod, claim.Name, err)
```
Comment (yuchengwu) [inline pkg/controller/statefulset/stateful_pod_control.go:311]: Emm.. it's just werid to have an `err` var as a method argument, why we need wrap this func body as `recordUnavailableClaimEvent`

File: pkg/controller/statefulset/stateful_set_control.go [hunk 5]
```diff
@@ -329,6 +330,11 @@ func (ssc *defaultStatefulSetControl) updateStatefulSet(
 		// If the ordinal could not be parsed (ord < 0), ignore the Pod.
 	}
 
+	err = ssc.resizePVCs(set, updateRevision, replicas)
+	if err != nil {
+		return nil, err
```
Comment (yuchengwu) [inline pkg/controller/statefulset/stateful_set_control.go:335]: not sure if ok to return nil `status`, does the status used by the caller ?

**Comment by yuchengwu** [comment]
Shall we respect the .spec.updateStrategy when resizing

**Comment by dims** [comment]
If you still need this PR then please rebase, if not, please close the PR

**Comment by dims** [comment]
This PR has the label work-in-progress, please revisit to see if you still need this, please close if not
